pkg/kstatus/observe/doc.go

```diff
@@ -0,0 +1,45 @@
+// Copyright 2020 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package observe is a library for computing the status of kubernetes resources
+// based on polling of resource state from a cluster. It can keep polling until
+// either some condition is met, or until it is cancelled through the provided
+// context. Updates on the status of resources are streamed back to the caller
+// through a channel.
+//
+// This package provides a simple interface based on the built-in
+// features. But the actual code is in the subpackages and there
+// are several interfaces that can be implemented to support custom
+// behavior.
+//
+// Note that while the observe package and its subpackages often use the
+// word Observer in both interfaces and objects, this does NOT refer to
+// the Observer design pattern. Rather it comes from k8s API conventions and
```

Comment:
I like the suggested name. It will require a broader change in the names across the library, so how about these names:
- Top level package (current name is `observe`): **poll** Seems a little too generic at first, but since it belongs inside the kstatus package, maybe it works.
- Main type for end users (current name is `StatusObserver`): **StatusPoller**
- Main function on the type (current name is `Observe`): **Poll**
- Return type of main function (currently `<-chan event.Event`): **<-chan event.Event** I think this works well, even with the new names. I think I prefer using context and go channel over `statusPoller.Cancel()`

I think I like that there is a separate subpackage that implements the main functionality and the package and type that provides the main API just wires things together. It allows users to compose things differently if they want to.
- Package for the core library functionality (currently `observer`): **poller** or **engine**? Not sure about either of them.
- Main type in the core library (currently `Observer`): **PollerEngine** I try to avoid using status as a prefix on every type. I think `engine` in the name hints that the package contains the core loop of the library.
- Main function on this type (currently `Observe`): **Poll**: I think it makes sense that the name here matches the function on the API of the outer package.

We also need a new name for the `ResourceObserver` interface and all the implementations. How about `StatusReader` for the interface and names like `DeploymentStatusReader` for the implementations?
I think the names for the two other key interfaces (`ClusterReader` and `StatusAggregator` works reasonably well).

WDYT?

Reply:
I think those names and that arrangement are great improvement and easier to grok.
Perfect.  If you want, we can submit this as is, and do a big name change acrsoss the board.